google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadFileChannel.java

```diff
@@ -0,0 +1,153 @@
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.channels.WritableByteChannel;
+
+class CloudStorageReadFileChannel extends FileChannel {
+  private static final String READ_ONLY = "This FileChannel is read-only";
+  private final SeekableByteChannel readChannel;
+
+  CloudStorageReadFileChannel(SeekableByteChannel readChannel) {
+    Preconditions.checkNotNull(readChannel);
+    this.readChannel = readChannel;
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException {
+    return readChannel.read(dst);
+  }
+
+  @Override
+  public long read(ByteBuffer[] dsts, int offset, int length) throws IOException {
+    long res = 0L;
+    synchronized (this) {
+      for (int i = offset; i < offset + length; i++) {
+        res += readChannel.read(dsts[i]);
+      }
+    }
+    return res;
+  }
+
+  @Override
+  public int write(ByteBuffer src) throws IOException {
+    throw new UnsupportedOperationException(READ_ONLY);
+  }
+
+  @Override
+  public long write(ByteBuffer[] srcs, int offset, int length) throws IOException {
+    throw new UnsupportedOperationException(READ_ONLY);
+  }
+
+  @Override
+  public long position() throws IOException {
+    return readChannel.position();
+  }
+
+  @Override
+  public FileChannel position(long newPosition) throws IOException {
+    readChannel.position(newPosition);
+    return this;
+  }
+
+  @Override
+  public long size() throws IOException {
+    return readChannel.size();
+  }
+
+  @Override
+  public FileChannel truncate(long size) throws IOException {
+    throw new UnsupportedOperationException(READ_ONLY);
+  }
+
+  @Override
+  public void force(boolean metaData) throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public long transferTo(long position, long count, WritableByteChannel target) throws IOException {
+    long res = 0L;
+    synchronized (this) {
+      long originalPosition = position();
+      position(position);
+      int blockSize = (int) Math.min(count, 0xfffffL);
+      int bytesRead = 0;
+      ByteBuffer buffer = ByteBuffer.allocate(blockSize);
+      while (res < count && bytesRead >= 0) {
+        buffer.position(0);
+        bytesRead = read(buffer);
+        if (bytesRead > 0) {
+          buffer.position(0);
+          buffer.limit(bytesRead);
+          target.write(buffer);
+          res += bytesRead;
+        }
+      }
+      position(originalPosition);
+    }
+    return res;
+  }
+
+  @Override
+  public long transferFrom(ReadableByteChannel src, long position, long count) throws IOException {
+    throw new UnsupportedOperationException(READ_ONLY);
+  }
+
+  @Override
+  public int read(ByteBuffer dst, long position) throws IOException {
+    synchronized (this) {
+      long originalPosition = position();
+      position(position);
```

Comment:
Isn't this redundant?

```
      long originalPosition = position();
      position(position);
```

Do we actually need `position(position);`?

Reply:
Yes, we need to position the channel where the read should begin. I've changed the name of the parameter to make it clearer.